include/boost/multiprecision/modular/modular_adaptor.hpp

```diff
@@ -0,0 +1,540 @@
+
+   modular_adaptor& operator=(modular_adaptor&& o)
+
+       BOOST_NOEXCEPT
+   {
+      m_base = std::move(o.base_data());
+      m_mod  = std::move(o.mod_data());
+      return *this;
+   }
+#endif
+
+   modular_adaptor& operator=(const char* s)
+   {
+      typedef typename mpl::front<unsigned_types>::type ui_type;
+      ui_type                                           zero = 0u;
+
+      using default_ops::eval_fpclassify;
+
+      if (s && (*s == '('))
+      {
+         std::string part;
+         const char* p = ++s;
+         while (*p && (*p != ',') && (*p != ')'))
+            ++p;
+         part.assign(s, p);
+         if (!part.empty())
+            m_base() = part.c_str();
+         else
+            m_base() = zero;
+         s = p;
+         if (*p && (*p != ')'))
+         {
+            ++p;
+            while (*p && (*p != ')'))
+               ++p;
+            part.assign(s + 1, p);
+         }
+         else
+            part.erase();
+         if (!part.empty())
+            m_mod() = part.c_str();
+         else
+            m_mod() = zero;
+      }
+      else
+      {
+         base_data() = s;
+         m_mod()     = zero;
+      }
+      return *this;
+   }
+
+   int compare(const modular_adaptor& o) const
+   {
+      // They are either equal or not:<
+      if (m_mod.compare(o.mod_data()) != 0)
+      {
+         BOOST_THROW_EXCEPTION(std::runtime_error("Could not compare modular number with different mod."));
+      }
+      Backend tmp1, tmp2;
+      mod_data().adjust_regular(tmp1, base_data());
+      mod_data().adjust_regular(tmp2, o.base_data());
+      return tmp1.compare(tmp2);
+   }
+
+   template <class T>
+   int compare(const T& val) const
+   {
+      using default_ops::eval_lt;
+      if (!eval_lt(m_mod, val))
+      {
+         BOOST_THROW_EXCEPTION(std::runtime_error("Could not compare modular number with mod bigger than compared number."));
+      }
+      Backend tmp;
+      mod_data().adjust_regular(tmp, base_data());
+      return tmp.compare(val);
+   }
+
+   inline void swap(modular_adaptor& o)
+   {
+      base_data().swap(o.base_data());
+      std::swap(mod_data(), o.mod_data());
+   }
+
+   inline std::string str(std::streamsize dig, std::ios_base::fmtflags f) const
+   {
+      Backend tmp;
+      mod_data().adjust_regular(tmp, base_data());
+      return tmp.str(dig, f);
+   }
+
+   inline void negate()
+   {
+      base_data().negate();
+      eval_add(base_data(), mod_data().get_mod().backend());
+   }
+
+   template <typename BackendT, expression_template_option ExpressionTemplates>
+   operator number<BackendT, ExpressionTemplates>()
+   {
+      return base_data();
+   };
+};
+
+template <class Result, class Backend>
+inline void eval_convert_to(Result* result, const modular_adaptor<Backend>& val)
+{
+   using default_ops::eval_convert_to;
+   eval_convert_to(result, val.base_data());
+}
+
+template <class Backend, class T>
+inline typename enable_if<is_arithmetic<T>, bool>
```

Comment:
Formatting SNAFU? ;)

Reply:
I've given up to clang-format. This is what it gave me. We can reformat this if required by disabling it with macro-attributes for these particular lines.